### ISSUE TITLE
[BUG] Possible bug in validation perplexity metric for DPO?

### ISSUE THREAD

**Issue by tmostak** [h2oai/h2o-llmstudio#594]
### 🐛 Bug

I'm not quite sure this is a bug, but after calibrating my hyperparameters for DPO (after running SFT on the model)  to get where validation loss falls through the training process, I am seeing validation perplexity increase, which I didn't think was possible (i.e validation loss and perplexity should be closely linked).

Relevant hyperparameters are Lora R: 64, Lora Alpha 256, Lora Dropout 0.05, Learning rate 0.00001, batch size 1, epochs 1, Gradient Clip 2, Beta 0.05. The dataset is composed of 620 SQL pairs that the model got wrong after SFT, with the corrected output.

Attaching a screenshot so you can see what I mean. Is this expected?

<img width="1538" alt="Screenshot 2024-01-30 at 11 45 47 AM" src="https://github.com/h2oai/h2o-llmstudio/assets/953210/c89dee1a-a1cb-437d-af44-79582ab52fca">

**Comment by maxjeblick** [comment]
Thanks a lot for posting, interesting findings!

I've checked Perplexity implementation, code-wise I haven't found any issues. 
For Dpo, validation loss and perplexity may exhibit different behavior, as dpo calculation is using
`policy_chosen_logps - policy_rejected_logps - (reference_chosen_logps - reference_rejected_logps)` 
whereas Perplexity is using `chosen_logits`, only. 

It is probably a good idea to add additional train/validation metrics (that will be logged to neptune) such as CE loss to better track the experiment.
I've created [this branch](https://github.com/h2oai/h2o-llmstudio/blob/max/dpo_ce_loss/llm_studio/src/models/text_dpo_modeling_model.py#L174) where `SampleAveragedCrossEntropyLoss` is used as a validation loss and ran some experiments on it. So far, val loss is in sync with Perplexity.
![Screenshot 2024-02-05 at 13-01-52 H2O LLM Studio](https://github.com/h2oai/h2o-llmstudio/assets/24281881/7679aa25-9048-4c35-a63a-0fa54f0d3708)

**Comment by maxjeblick** [comment]
I digged a bit deeper into the loss curves in [this branch](https://github.com/h2oai/h2o-llmstudio/tree/max/log_more_dpo_metrics); cross entropy is logged for both accepted as well as rejected samples; alongside with the corresponding perplexity. 

Regarding the different behavior of loss vs. perplexity, my explanation for this is that evaluation samples the model struggles to predict can disproportionately impact the overall perplexity.


As an example, suppose we have a dataset with 4 samples where the third sample is out-of-distribution of the SFT model. 
-  After **epoch 1**, suppose the model has the following cross entropy loss for each sample: `(1, 1, 9, 1)` (third answer is hard to predict).
   **Mean cross-entropy** (cross-entropy per sample / num_samples) will be `12/4 = 3`
   **Mean perplexity**  (perplexity per sample / num_samples) will be `mean([exp(1), exp(1), exp(9), exp(1)]) ~ 2027.8`.
- After **epoch 2,** suppose the model has the following cross entropy loss for each sample: `(4, 4, 4, 4)` (model adapts to third sample at the cost of predicting the other samples worse)
    **Mean cross-entropy** will be `16/4 = 4`
    **Mean perplexity** will be `exp(4) / 4 ~ 13.6`

My assumption is that this explains the behavior observed (and it is not caused by a bug).
